src/main/java/org/springframework/data/r2dbc/repository/query/ConditionFactory.java

```diff
@@ -0,0 +1,307 @@
+                        ? NotLike.create(lhs, rhs)
+                        : Conditions.like(lhs, rhs);
+            }
+            case TRUE:
+            case FALSE: {
+                Expression pathExpression = createPropertyPathExpression(part.getProperty());
+                // TODO: include factory methods for '= TRUE/FALSE' conditions into spring-data-relational
+                return Conditions.isEqual(pathExpression,
+                        SQL.literalOf((Object) (part.getType() == Part.Type.TRUE ? "TRUE" : "FALSE")));
+            }
+            case SIMPLE_PROPERTY: {
+                Expression pathExpression = createPropertyPathExpression(part.getProperty());
+                ParameterMetadata parameterMetadata = parameterMetadataProvider.next(part);
+                if (parameterMetadata.isIsNullParameter()) {
+                    return Conditions.isNull(pathExpression);
+                }
+
+                BindMarker bindMarker = createBindMarker(parameterMetadata);
+                Expression lhs = upperIfIgnoreCase(part, pathExpression);
+                Expression rhs = upperIfIgnoreCase(part, bindMarker, parameterMetadata.getType());
+                return Conditions.isEqual(lhs, rhs);
+            }
+            case NEGATING_SIMPLE_PROPERTY: {
+                Expression pathExpression = createPropertyPathExpression(part.getProperty());
+                ParameterMetadata parameterMetadata = parameterMetadataProvider.next(part);
+                BindMarker bindMarker = createBindMarker(parameterMetadata);
+                Expression lhs = upperIfIgnoreCase(part, pathExpression);
+                Expression rhs = upperIfIgnoreCase(part, bindMarker, parameterMetadata.getType());
+                return Conditions.isEqual(lhs, rhs).not();
+            }
+            default:
+                throw new IllegalArgumentException("Unsupported keyword " + type);
+        }
+
+    }
+
+    @NotNull
+    private Expression createPropertyPathExpression(PropertyPath propertyPath) {
+        RelationalPersistentEntity<?> entity = getPersistentEntity(propertyPath);
+        RelationalPersistentProperty property = entity.getRequiredPersistentProperty(propertyPath.getSegment());
+        Table table = SQL.table(dataAccessStrategy.toSql(entity.getTableName()));
+        return SQL.column(dataAccessStrategy.toSql(property.getColumnName()), table);
+    }
+
+    @NotNull
+    @SuppressWarnings("unchecked")
+    private RelationalPersistentEntity<?> getPersistentEntity(PropertyPath propertyPath) {
+        R2dbcConverter converter = dataAccessStrategy.getConverter();
+        return converter.getMappingContext().getRequiredPersistentEntity(propertyPath.getOwningType());
+    }
+
+    @NotNull
+    private BindMarker createBindMarker(ParameterMetadata parameterMetadata) {
+        if (parameterMetadata.getName() != null) {
+            return SQL.bindMarker(":" + parameterMetadata.getName());
```

Comment:
Using named parameters manifests the requirement that named parameter processing is enabled in `DatabaseClient`. We tried to not stick with such an assumption. Ideally, we generate directly SQL that does not require post-processing by accessing `Dialect` when generating SQL to use the proper `BindMarkersFactory`.

Reply:
I can suggest the following soultion:

1. at first we should expose `Dialect` because we'll need it to create correct bind markers; in my opinion `ReactiveDataAccessStrategy` interface is a good candidate (if so I'm not sure whether default method is required since this interface has only one implementation);
2. pass `Dialect` to the constructor of `ParameterMetadataProvider`
```
public class PartTreeR2dbcQuery extends AbstractR2dbcQuery {
    ...
    protected BindableQuery createQuery(RelationalParameterAccessor accessor) {
        ...
        ParameterMetadataProvider parameterMetadataProvider = new ParameterMetadataProvider(accessor, likeEscaper,
				dataAccessStrategy.getDialect());
...
```
3. in `ParameterMetadataProvider` create `BindMarkers`  using current `BindMarkersFactory`
```
class ParameterMetadataProvider {
    ...
    private ParameterMetadataProvider(Parameters<?, ?> parameters,
            @Nullable Iterator<Object> bindableParameterValueIterator, LikeEscaper likeEscaper, R2dbcDialect dialect) {
        ...
        this.bindMarkers = dialect.getBindMarkersFactory().create();
    }
...
```
4. save parameter placeholder in `ParameterMetadata`
```
class ParameterMetadataProvider {
    ...
    public ParameterMetadata next(Part part) {
        ...
        ParameterMetadata metadata = ParameterMetadata.builder()
                .type(parameter.getType())
                .partType(part.getType())
                .name(parameterName)
                .isNullParameter(getParameterValue() == null && Part.Type.SIMPLE_PROPERTY.equals(part.getType()))
                .placeholder(getPlaceholder(parameterName))
                .likeEscaper(likeEscaper)
                .build();
        ...
    }

    private String getPlaceholder(String parameterName) {
        return bindMarkers.next(parameterName).getPlaceholder();
    }
...
```
5. use placeholder to create bind marker
```
class ConditionFactory {
...
    @NonNull
    private BindMarker createBindMarker(ParameterMetadata parameterMetadata) {
        return SQL.bindMarker(parameterMetadata.getPlaceholder());
    }
...
```
6. always use placeholder to bind parameter value
```
class PartTreeBindableQuery implements BindableQuery {
...
    public <T extends DatabaseClient.BindSpec<T>> T bind(T bindSpec) {
        T bindSpecToUse = bindSpec;
        int index = 0;

        for (Object value : accessor.getValues()) {
            ParameterMetadata metadata = parameterMetadataProvider.getParameterMetadata(index++);
            Class<?> parameterType = metadata.getType();
            if (value == null) {
                bindSpecToUse = bindSpecToUse.bindNull(metadata.getPlaceholder(), parameterType);
            } else {
                bindSpecToUse = bindSpecToUse.bind(metadata.getPlaceholder(), metadata.prepare(value));
            }
        }
        return bindSpecToUse;
    }
...
```

Reply:
I've just realized that my approach won't work for databases where question mark is used as a placeholder (MySQL for example). Maybe during parameter value binding I should check whether placeholder is '?' and use binding by index if it is true.